### PR TITLE
Appmenu folders for organizing qubes

### PR DESCRIPTION
**Pull request by amrosia**
### Summary
When users have many qubes, the Applications pane becomes difficult to scan and navigate quickly. Related qubes are mixed in one long list, which makes common workflows slower.

This change adds folder-based organization to the VM list in the appmenu, so users can group related qubes and reduce visual clutter.

### What this PR adds
- Folder grouping for qubes in the Applications pane
- Collapsible folder sections for faster navigation
- Per-scope organization (Apps/Templates/Service each keep independent folder layout)
- Folder management actions from context menus (move qube, create folder, rename/delete folder, reorder folders)
- Folder ordering controls (move up/down)

### Persistence model
Appmenu folder state is stored in qvm-features:
- VM-level feature keys store per-qube folder assignment
- Behavior remains backward-compatible for users without folder features configured.

Warning: AI-generated code
This change set was written with AI, then tested manually.

Screenshot_20260403_104717
Screenshot_20260403_105005
Screenshot_20260403_105120

### CODE REVIEW

**Comment by marmarta** [comment]
I think this is promising, but  tests fail (the fail on desktop-linux-manager is known, but there is a bunch of fails for menu tests too).

**Comment by amrosia** [comment]
Thanks for the review. I fixed the menu test failures.

In my tests, the result was: 31 passed, 1 skipped, 7 warnings.

I also added 5 tests for folder functionality in test_app_page.py and test_vmmanager.py

Coverage includes create/assign/rename/delete, ordering and collapse behavior, scope-specific state handling, submenu entries, and VM feature-event refresh behavior.

**Comment by marmarta** [comment]
PipelineRetryFailed

**Comment by marmarek** [comment]
PipelineRetry

**Comment by ben-grande** [comment]
Could you please take a look at the errors: https://gitlab.com/QubesOS/qubes-desktop-linux-menu/-/jobs/14220167274

```
************* Module qubes_menu.custom_widgets
qubes_menu/custom_widgets.py:442:4: C0116: Missing function or method docstring (missing-function-docstring)
************* Module qubes_menu.vm_manager
qubes_menu/vm_manager.py:347:38: W0212: Access to a protected member _safe_feature_get of a client class (protected-access)
```

**Comment by ben-grande** [comment]
Some if not all of the black formatter issues will be fixed in https://github.com/QubesOS/qubes-desktop-linux-menu/pull/74. I didn't compare the files to see if there were things you introduced.

**Comment by amrosia** [comment]
Understood, in this case I've only run black on files that my commits introduced.

**Comment by marmarta** [comment]
I have a couple questions:
- how does this interact with sorting? with sorting running qubes to top?
- I'm not sure about showing an 'Ungrouped' section by default if there are no folders
- how persistent are the visual settings (collapsing/expanding/order)?

File: qubes_menu/constants.py [hunk 1]
```diff
@@ -39,6 +39,18 @@
...
+FOLDER_FEATURE_TEMPLATES = "menu-folder-templates"
+FOLDER_FEATURE_SERVICE = "menu-folder-service"
+FOLDERS_FEATURE = "menu-folders"
+FOLDERS_COLLAPSED_FEATURE = "menu-folders-collapsed"
+FOLDERS_FEATURE_APPS = "menu-folders-apps"
+FOLDERS_FEATURE_TEMPLATES = "menu-folders-templates"
+FOLDERS_FEATURE_SERVICE = "menu-folders-service"
+FOLDERS_COLLAPSED_FEATURE_APPS = "menu-folders-collapsed-apps"
+FOLDERS_COLLAPSED_FEATURE_TEMPLATES = "menu-folders-collapsed-templates"
+FOLDERS_COLLAPSED_FEATURE_SERVICE = "menu-folders-collapsed-service"
```
Comment (ben-grande) [inline qubes_menu/constants.py]: There are too many features. It would also require adding new features if new tabs are added. The separation of `APPS`, `TEMPLATES`, `SERVICE` is already done by the menu, thus there doesn't seem to be a need for features specialized for type.

Also, we can't have one feature named `menu-folder` and another one named `menu-folders`, it is easy to mistake one for the other.

Can we have just:

- `menu-folder`
- `menu-folder-collapsed`

**Comment by amrosia** [comment]
> I have a couple questions:
> 
>     * how does this interact with sorting? with sorting running qubes to top?
> 
>     * I'm not sure about showing an 'Ungrouped' section by default if there are no folders
> 
>     * how persistent are the visual settings (collapsing/expanding/order)?

1. Sorting running VMs to top works by scoping it per-folder. A running VM will float to the top of its own folder section, but won't jump ahead of VMs in other folders. The sort key at _row_sort_key() uses folder_index as the primary dimension, then state_prefix (0 for Running, 1 otherwise) within that folder.
https://github.com/QubesOS/qubes-desktop-linux-menu/blob/59c0d3a8b8e642c4d66f7122fc42622b50c23917/qubes_menu/application_page.py#L273-L290

2. Good point. I've made a commit that will not display Ungrouped unless there is at least 1 folder created by the user. The check is in _rebuild_folder_rows(): if filtering out "Ungrouped" from folder_order leaves nothing, no folder rows are created at all and VMs are rendered flat.
https://github.com/QubesOS/qubes-desktop-linux-menu/blob/59c0d3a8b8e642c4d66f7122fc42622b50c23917/qubes_menu/application_page.py#L756-L763

Though there is one issue: when a qube is moved from a folder where this qube was the only one, the folder will remain empty. With the current code, empty folders are hidden,
https://github.com/QubesOS/qubes-desktop-linux-menu/blob/59c0d3a8b8e642c4d66f7122fc42622b50c23917/qubes_menu/application_page.py#L780-L787
but this might cause confusion as to why "Ungrouped" is still displayed, because even though folders exist, they are not visible because they contain no qubes. I thought that automatically removing empty folders would make sense ergonomically, but I'm cautious about adding it since it's implicit, so I'd like to hear your feedback on this one before I add this.

3. All persisted in dom0's Qubes features. Folder order and collapsed state are saved per-scope (Apps/Templates/Service VMs) on dom0 via _save_folder_state() and _save_collapsed_state()
https://github.com/QubesOS/qubes-desktop-linux-menu/blob/59c0d3a8b8e642c4d66f7122fc42622b50c23917/qubes_menu/application_page.py#L736-L747

Per-VM folder assignments are saved as a feature on each VM, set in _assign_folder().
Because they are stored in features, they survive menu restarts and reboots.
https://github.com/QubesOS/qubes-desktop-linux-menu/blob/59c0d3a8b8e642c4d66f7122fc42622b50c23917/qubes_menu/application_page.py#L508-L526

File: qubes_menu/constants.py (hunk 1 above)
Comment (amrosia) [reply qubes_menu/constants.py]: I agree.

In case of AppVMs that are also marked as Disposable templates, it's one qube
that appears in both `Apps` and `Templates` simultaneously, and I suppose it
would be flexible to let those qubes be in 2 groups, 1 for `Apps` and second
for `Templates`.

So I've made menu-folder be a JSON dict that can store more than 1 scope so
that AppVM marked as disposable templates could have independent folder assignments per category.

Examples:

in dom0:
menu-folder-collapsed              {"apps": {"folders": ["Ungrouped", "apptesting"], "collapsed": []}, "templates": {"folders": ["Ungrouped", "tpltesting"], "collapsed": ["tpltesting"]}, "service": {"folders": ["Ungrouped", "net"], "collapsed": ["net"]}}

in an AppVM where template_for_dispvms=True:
menu-folder                {"apps": "apptesting", "templates": "tpltesting"}

in a normal AppVM assigned to "apptesting" folder:
menu-folder                {"apps": "apptesting"}

in a normal AppVM with no folder assigned:
(no feature)

Those examples use "apptesting", "tpltesting" and "net" as folder names.

Comment (ben-grande) [reply qubes_menu/constants.py]: Currently, the same menu-items you see on "APPS" and "TEMPLATES" of a disposable template, are the same, later a new feature could be made to show different set of applications depending on the tab.

I don't think the folders should be responsible for that. Can we limit folder to be only section separator? This would avoid JSON, as the value would simply be a list of strings space separated, like `menu-items`.

If you notice other features, they are human readable and writable, JSON is not as easy to read as just text.

Comment (marmarta) [reply qubes_menu/constants.py]: JSON is not terrible for readability compared to some other formats, but I agree that this might be a bit too complex; to be honest I'm torn between storing this in qubes features or in some other format like Gtk's settings. Another idea that comes to my mind is making a per-qube feature that contains folders a given qube is in, instead of a big dom0 feature. This would interact more nicely with backups - you restore your qube and it is back in the right folder. I'm not sure if it's that much worth it to spend a lot of time making sure you can have different folders for "run new disposable from this disposable template" and "run within this disposable template", but I am very convinceable here.

**Comment by marmarta** [comment]
> 1. Sorting running VMs to top works by scoping it per-folder. A running VM will float to the top of its own folder section, but won't jump ahead of VMs in other folders. The sort key at _row_sort_key() uses folder_index as the primary dimension, then state_prefix (0 for Running, 1 otherwise) within that folder.

awesome, that's ideal

2. I'm not sure about hiding empty folders - I imagine that when moving things around, a folder might be empty for a moment? OTOH using the "qube has a feature saying what menu folders it is in" idea above, this would potentially be a moot issue?

3. ok, thanks!

**Comment by marmarta** [comment]
BTW, the more I look at it the more I like it :) I will do more intense testing once the heatwave passes and I'm not scared to melt my testing machine.

**Comment by amrosia** [comment]
1. I rebased the branch onto the current upstream main to resolve the merge conflict with #74.
2. I've also simplified the feature set to only `menu-folder-order` (dom0), `menu-folder-collapsed` (dom0), and `menu-folder` (per qube). I think features fit better here and are consistent with how the rest of the app menu is implemented. They use space-separated string lists, as ben-grande suggested, and I've added escaping for folder names that contain spaces. Overall, this simplification means a qube can only belong to one folder at a time, but after my testing I think usability doesn't suffer from it, and it works out fine.
3. Empty folders cannot exist by definition: a folder only exists while at least one qube references it, so no empty section is ever shown.
4. Folder ordering is global, independent of which qubes are visible in the current tab, so "Move folder up/down" swaps the folders' positions in that single global list. This way the ordering fits into one feature.
5. Because folder assignments are stored per qube, they should survive qube backup and restore.
6. This is still very experimental, and I think more testing is required to catch edge cases I might have missed.

**Comment by ben-grande** [comment]
> 2\. I've also simplified the feature set to only `menu-folder-order` (dom0), `menu-folder-collapsed` (dom0), and `menu-folder` (per qube).

Much better. About referring to them as `dom0`, it is partially true, but we should refer to it as "interface qube". I see you used `local_name`, which is correct, it will get the name of the qube running core-admin-client.

- https://doc.qubes-os.org/en/latest/user/reference/glossary.html#term-interface-qube

So on the comments and variable names you made calling it dom0, please change it to something else, such as `guivm` for code or `interface qube` for pretty name.

**Comment by amrosia** [comment]
Fixed terminology.

File: qubes_menu/custom_widgets.py [hunk 2]
```diff
@@ -392,6 +407,64 @@ def sort_order(self):
...
+    def __init__(
+        self,
+        folder_name: str,
+        collapsed: bool = False,
+        toggle_handler: Optional[Callable] = None,
+        menu_handler: Optional[Callable] = None,
+    ):
+        super().__init__()
+        self.folder_name = folder_name
+        self.vm_name = folder_name
```
Comment (marmarta) [inline qubes_menu/custom_widgets.py:422]: This is confusing. Needs at least a comment why a folder has vm_name property, but I think this shouldn't be here and if somewhere in the code this is relied upon, there's a mistake there.

File: qubes_menu/custom_widgets.py [hunk 3]
```diff
@@ -392,6 +407,64 @@ def sort_order(self):
...
+        """Update the collapse arrow and folder label."""
+        arrow = "▸" if self.collapsed else "▾"
+        self.arrow.set_text(arrow)
+        escaped_folder = GLib.markup_escape_text(self.folder_name)
+        self.label.set_markup(f"<b>{escaped_folder}</b>")
+        self.main_box.show_all()
+
+    def _on_button_press(self, _widget, event):
+        """Handle left click for collapse toggling and right click menu."""
+        if event.button == 3 and self.menu_handler:
```
Comment (marmarta) [inline qubes_menu/custom_widgets.py:459]: use named constants, not hardcoded numbers.

File: qubes_menu/application_page.py [hunk 4]
```diff
@@ -229,21 +251,102 @@ def _on_map_vm_list(self, *_args):
...
-                "0"
-                if (other_entry.vm_entry.power_state == "Running")
-                else "1" + other_sort_name
-            )
-        return my_sort_name > other_sort_name
+    def _row_sort_key(self, row):
+        if isinstance(row, FolderRow):
+            return row.sort_order
+
+        if not isinstance(row, VMRow):
```
Comment (marmarta) [inline qubes_menu/application_page.py:258]: this should not happen? if it happens, there's something subtly wrong with the code...

File: qubes_menu/application_page.py [hunk 5]
```diff
@@ -229,21 +251,102 @@ def _on_map_vm_list(self, *_args):
...
+        A single plain-text folder name is stored per VM; the same folder
+        applies in every menu scope (Apps, Templates, Service) in which the
+        VM is shown.
+        """
+        return self._safe_feature_get(vm_entry.vm, constants.FOLDER_FEATURE, "")
+
+    def _effective_vm_folder(self, vm_entry: VMEntry) -> str:
+        """Resolve the VM folder name to an existing folder.
+
+        Unknown/missing folder names are treated as Ungrouped to avoid
```
Comment (marmarta) [inline qubes_menu/application_page.py:290]: This looks like restoring VM from backup will result in Ungrouped VMs, unless the user manually recreates the folder. Folder should be restored when VM is restored.

File: qubes_menu/application_page.py [hunk 6]
```diff
@@ -229,21 +251,102 @@ def _on_map_vm_list(self, *_args):
...
+        if not folder_name:
+            return self.UNGROUPED
+        if folder_name not in self.folder_order:
+            return self.UNGROUPED
+        return folder_name
+
+    @staticmethod
+    def _set_vm_folder(vm_entry: VMEntry, folder_name: str):
+        """Set *folder_name* on *vm_entry* or clear the assignment."""
+        try:
```
Comment (marmarta) [inline qubes_menu/application_page.py:304]: handle what happens if user has no permissions to set the feature

File: qubes_menu/application_page.py [hunk 7]
```diff
@@ -229,21 +251,102 @@ def _on_map_vm_list(self, *_args):
...
+                used.add(folder)
+        return used
+
+    def _order_folders(self, folders: List[str], used: Set[str]) -> List[str]:
+        """Stored order filtered to the folders actually in use.
+
+        Names no qube is assigned to disappear, so empty sections are never
+        shown. Ungrouped is the virtual fallback bucket: it is never removed
+        and keeps the user-placed position (first on a fresh install); no
+        folder is ever pinned to an index.
```
Comment (marmarta) [inline qubes_menu/application_page.py:327]: That's confusing. What is pinning a folder to an index?  What is "stored older filtered to the folders actually in use"?

File: qubes_menu/application_page.py [hunk 8]
```diff
@@ -229,21 +251,102 @@ def _on_map_vm_list(self, *_args):
...
+        """
+        result: List[str] = []
+        for name in folders:
+            if name == self.UNGROUPED:
+                if self.UNGROUPED not in result:
+                    result.append(self.UNGROUPED)
```
Comment (marmarta) [inline qubes_menu/application_page.py:333]: that looks pointless, judging from line 340

File: qubes_menu/application_page.py [hunk 9]
```diff
@@ -229,21 +251,102 @@ def _on_map_vm_list(self, *_args):
...
+        """Return the set of folder names actually assigned to VMs."""
+        used: Set[str] = set()
+        for vm_entry in vm_entries:
+            folder = self._vm_folder(vm_entry)
+            if folder and folder != self.UNGROUPED:
+                used.add(folder)
+        return used
+
+    def _order_folders(self, folders: List[str], used: Set[str]) -> List[str]:
+        """Stored order filtered to the folders actually in use.
```
Comment (marmarta) [inline qubes_menu/application_page.py:322]: That is not what the function does, as far as I can understand the grammar here.

File: qubes_menu/application_page.py [hunk 10]
```diff
@@ -229,21 +251,102 @@ def _on_map_vm_list(self, *_args):
...
+                result.append(name)
+        for name in sorted(used):
+            if name not in result:
+                result.append(name)
+        if self.UNGROUPED not in result:
+            result.insert(0, self.UNGROUPED)
+        return result
+
+    @staticmethod
+    def _safe_feature_get(vm, feature_name: str, default=""):
```
Comment (marmarta) [inline qubes_menu/application_page.py:345]: why this? why all the stripping? plus, don't use broad except here, catch actual exceptions

File: qubes_menu/application_page.py [hunk 11]
```diff
@@ -338,6 +910,7 @@ def _app_clicked(self, _widget: Gtk.Widget, row: AppEntry):
     def _button_toggled(self, widget: Gtk.ToggleButton):
         if not widget.get_active():
             return
+        self._rebuild_folder_rows()
```
Comment (marmarta) [inline qubes_menu/application_page.py:913]: absolutely not needed. Folders should not be recreated on every button toggle.

File: qubes_menu/application_page.py [hunk 12]
```diff
@@ -283,16 +386,485 @@ def _vm_callback(self, vm_entry: VMEntry):
...
+            return False
+
+        vm_folder = self._effective_vm_folder(row.vm_entry)
+        if not self.toggle_buttons.filter_function(row):
+            return False
+        if vm_folder in self.collapsed_folders:
+            return False
+        return True
+
+    def _prompt_for_text(self, title: str, initial_value: str = ""):
```
Comment (marmarta) [inline qubes_menu/application_page.py:830]: I think it might be a good idea to reuse some code from desktop-linux-manager here, because as it stands, we reinvent messagebox every time.

File: qubes_menu/application_page.py [hunk 13]
```diff
@@ -283,16 +386,485 @@ def _vm_callback(self, vm_entry: VMEntry):
...
+            return None
+        index = self.folder_order.index(folder_name) + direction
+        while 0 <= index < len(self.folder_order):
+            candidate = self.folder_order[index]
+            if self._folder_has_visible_vms(candidate):
+                return candidate
+            index += direction
+        return None
+
+    def _move_folder(self, _widget, folder_name: str, direction: int):
```
Comment (marmarta) [inline qubes_menu/application_page.py:658]: the moving structure is really overengineered. Not sure there is a need for maintaining sort order for unused folders?

File: qubes_menu/application_page.py [hunk 14]
```diff
@@ -283,16 +386,485 @@ def _vm_callback(self, vm_entry: VMEntry):
...
+        else:
+            self.collapsed_folders.add(folder_row.folder_name)
+            folder_row.collapsed = True
+
+        folder_row.update_contents()
+        self._save_folder_state()
+        self.vm_list.invalidate_filter()
+
+    def _show_folder_row_menu(self, row: FolderRow, event):
+        if event.button != 3:
```
Comment (marmarta) [inline qubes_menu/application_page.py:580]: use named Gtk constants.

File: qubes_menu/application_page.py [hunk 15]
```diff
@@ -283,16 +386,485 @@ def _vm_callback(self, vm_entry: VMEntry):
...
+    def _create_folder_for_vm(self, _widget, vm_entry: VMEntry):
+        folder_name = self._prompt_for_text("Create folder")
+        if folder_name is None:
+            return
+        folder_name = folder_name.strip()
+        if not folder_name:
+            return
+        self._assign_folder(vm_entry, folder_name)
+
+    def _assign_folder_to_vm(
```
Comment (marmarta) [inline qubes_menu/application_page.py:489]: this is unneeded, add a parameter to assign folder if you want to directly use it from event

File: qubes_menu/application_page.py [hunk 16]
```diff
@@ -283,16 +386,485 @@ def _vm_callback(self, vm_entry: VMEntry):
...
+        dialog.show_all()
+        response = dialog.run()
+        text = entry.get_text()
+        dialog.destroy()
+
+        if response != Gtk.ResponseType.OK:
+            return None
+        return text
+
+    def _confirm(self, title: str, text: str) -> bool:
```
Comment (marmarta) [inline qubes_menu/application_page.py:855]: as above

File: qubes_menu/custom_widgets.py [hunk 17]
```diff
@@ -324,13 +329,22 @@ def __init__(
...
+
         self.update_contents(
             update_power_state=True,
             update_label=True,
             update_has_network=True,
             update_type=True,
         )
 
+    def _on_button_press(self, _widget, event):
+        if event.button == 3 and self.folder_menu_handler:
```
Comment (marmarta) [inline qubes_menu/custom_widgets.py:343]: as above

File: qubes_menu/custom_widgets.py [hunk 18]
```diff
@@ -324,13 +329,22 @@ def __init__(
...
+            self.folder_menu_handler(self, event)
+            return True
```
Comment (marmarta) [inline qubes_menu/custom_widgets.py:345]: add a comment explaining why return True/False, this is later difficult to remember when maintaining the code and can lead to subtle bugs slipping through review

File: qubes_menu/custom_widgets.py [hunk 19]
```diff
@@ -392,6 +407,64 @@ def sort_order(self):
...
+        self.main_box.show_all()
+
+    def _on_button_press(self, _widget, event):
+        """Handle left click for collapse toggling and right click menu."""
+        if event.button == 3 and self.menu_handler:
+            self.menu_handler(self, event)
+            return True
+        if event.button == 1 and self.toggle_handler:
+            self.toggle_handler(self)
+            return True
```
Comment (marmarta) [inline qubes_menu/custom_widgets.py:464]: as above, explain why True/False so that future maintainer doesn't have to check docs every time

File: qubes_menu/vm_manager.py [hunk 20]
```diff
@@ -29,6 +29,22 @@
 from . import constants
 
 
+def _to_bool(value) -> bool:
```
Comment (marmarta) [inline qubes_menu/vm_manager.py:32]: why is this needed?

File: qubes_menu/vm_manager.py [hunk 21]
```diff
@@ -43,36 +59,70 @@ def __init__(self, vm: QubesVM):
...
+        self.sort_name = ""
+        self._update_sort_name()
 
         try:
-            self._internal = bool(
+            self._internal = _to_bool(
                 self.vm.features.check_with_template("internal", False)
             )
-        except qubesadmin.exc.QubesDaemonAccessError:
+        except Exception:  # pylint: disable=broad-except
```
Comment (marmarta) [inline qubes_menu/vm_manager.py:73]: this and everywhere else: do not do this unless very well justified, and it is not justified here. In general, I don't see why so much rewrite of the code here to make it more prone to invisible failures?

File: qubes_menu/vm_manager.py [hunk 22]
```diff
@@ -43,36 +59,70 @@ def __init__(self, vm: QubesVM):
...
+            )
+        except Exception:  # pylint: disable=broad-except
+            self.show_dispvm_template_in_apps = False
         self.entries: List = []
 
+    def safe_feature_get(self, feature_name: str, default=""):
+        """Return the named feature value for this VM, or default if unset."""
+        try:
+            return str(self.vm.features.get(feature_name, default)).strip()
+        except Exception:  # pylint: disable=broad-except
```
Comment (marmarta) [inline qubes_menu/vm_manager.py:113]: as above.

File: qubes_menu/vm_manager.py [hunk 23]
```diff
@@ -226,20 +287,29 @@ def register_new_vm_callback(self, func):
...
         try:
             vm: QubesVM = self.qapp.domains[vm_name]
         except KeyError:
             return None
         try:
-            if vm.features.check_with_template("internal", False):
+            if _to_bool(vm.features.check_with_template("internal", False)):
                 return None
-        except qubesadmin.exc.QubesDaemonAccessError:
+        except Exception:  # pylint: disable=broad-except
```
Comment (marmarta) [inline qubes_menu/vm_manager.py:301]: in general, please remove all superflouos changes to base code that are not justified by needs of this PR before submitting the PR.

File: qubes_menu/vm_manager.py [hunk 24]
```diff
@@ -226,20 +287,29 @@ def register_new_vm_callback(self, func):
...
             pass
 
         return self._add_vm(vm)
 
+    @staticmethod
+    def _vm_name(vm_or_name) -> str:
```
Comment (marmarta) [inline qubes_menu/vm_manager.py:307]: this is very confusing, I don't see how this is needed

File: qubes_menu/vm_manager.py [hunk 25]
```diff
@@ -253,10 +323,23 @@ def _add_vm(self, vm) -> Optional[VMEntry]:
         return entry
 
     def _add_domain(self, _submitter, _event, vm, **_kwargs):
-        self.load_vm_from_name(vm)
+        if isinstance(vm, str):
+            self.load_vm_from_name(vm)
+            return
+
+        vm_name = self._vm_name(vm)
```
Comment (marmarta) [inline qubes_menu/vm_manager.py:330]: why those changes?

File: qubes_menu/vm_manager.py [hunk 26]
```diff
@@ -253,10 +323,23 @@ def _add_vm(self, vm) -> Optional[VMEntry]:
...
+
+        if self._add_vm(vm):
+            return
+
+        # fallback path in case event payload object is incomplete
+        self.load_vm_from_name(vm_name)
 
     def _remove_domain(self, _submitter, _event, vm, **_kwargs):
-        vm_entry = self.vms.get(vm)
+        vm_name = self._vm_name(vm)
```
Comment (marmarta) [inline qubes_menu/vm_manager.py:341]: why? again, please do not submit random other changes, this is not relevant to the PR.

File: qubes_menu/vm_manager.py [hunk 27]
```diff
@@ -307,12 +390,21 @@ def _update_domain_feature(
         if not vm_entry:
             return
 
-        if value == "False":
-            value = False
-        value = bool(value)
-
         try:
             if feature == "internal":
+                if "delete" in str(_event):
```
Comment (marmarta) [inline qubes_menu/vm_manager.py:395]: as above.

File: qubes_menu/vm_manager.py [hunk 28]
```diff
@@ -321,9 +413,21 @@ def _update_domain_feature(
                     if derived_vm_entry:
                         derived_vm_entry.internal = value
             if feature == "servicevm":
-                vm_entry.service_vm = value
+                vm_entry.service_vm = _to_bool(
```
Comment (marmarta) [inline qubes_menu/vm_manager.py:416]: as above

**Comment by marmarta** [comment]
I've been testing this (and I like it in general quite much, at least the usability side, not much of a fan of the code, it's not really streamlined/maintainable), I have comments and edge cases:
- a new disposable is always Ungrouped, even if its parent is in a group (I think it should land in the parent folder), and this makes for very weird graphical anomalies when the "running qubes to top" option is disabled
- new folder lands at the bottom, which is unintuitive for me; I would expect it to land above Ungrouped (in general, I'd expect Ungrouped to stay on bottom)
